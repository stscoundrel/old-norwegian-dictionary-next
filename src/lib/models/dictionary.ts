import { DictionaryEntry as RawDictionaryEntry } from 'old-norwegian-dictionary/dist'

export type OriginalDictionaryEntry = RawDictionaryEntry

export interface DictionaryEntry extends RawDictionaryEntry {
    slug: string,
}
