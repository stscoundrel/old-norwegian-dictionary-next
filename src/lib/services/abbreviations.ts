import { findAbbreviations } from 'old-norwegian-dictionary-abbreviations'
import { abbreviate } from 'abbreviatrix'

export interface Abbreviation{
    abbreviation: string,
    explanation: string
}

export const getAbbreviations = ({ definition }): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  const abbreviations = findAbbreviations(definition)
  abbreviations.forEach((explanation, abbreviation) => {
    if (!abbreviationSet.has(abbreviation)) {
      abbreviationSet.add(abbreviation)
      combinedAbbreviations.push({ abbreviation, explanation })
    }
  })

  return combinedAbbreviations;
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations: Abbreviation[],
): string => {
  let result = content

  abbreviations.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}
