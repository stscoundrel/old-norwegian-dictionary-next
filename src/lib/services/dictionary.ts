import { getDictionary } from 'old-norwegian-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { slugifyWord, slugifyLetter } from '../utils/slugs'
import { OriginalDictionaryEntry, DictionaryEntry } from '../models/dictionary'

export interface AlphabetLetter{
  letter: string,
  slug: string,
}

const addSlugs = (words: OriginalDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  return formattedWords
}

export const getByLetter = (letter: string): DictionaryEntry[] => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === word)[0]
)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}
