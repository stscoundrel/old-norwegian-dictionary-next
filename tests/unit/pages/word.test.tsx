import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import { DictionarySource } from 'scandinavian-dictionary-crosslinker'
import Word, { getStaticProps, getStaticPaths } from 'pages/word/[word]'
import { getAlphabet } from 'lib/services/dictionary'

const mockHandler = jest.fn()

/**
 * Mock router
 */
jest.mock('next/router', () => ({
  useRouter() {
    return {
      locale: undefined,
      defaultLocale: undefined,
      asPath: '/test',
      back: mockHandler,
    }
  },
}))

describe('Word page: render & usage', () => {
  const word = {
    word: 'þrályndi',
    partOfSpeech: 'n',
    definition: 'þrályndi, n. d. s.; hélt honum aptr fornsiðvenja ok mykyt þrályndi Barl. 12524;Fgrdl. 5816.',
    slug: 'thralyndi-2',
  }

  const abbreviations = [
    {
      abbreviation: 'f.',
      explanation: 'Feminin.',
    },
    {
      abbreviation: 'L.',
      explanation: 'Linje.',
    },
    {
      abbreviation: 'm.',
      explanation: 'Masculin.',
    },
  ]

  const crosslinks = [
    {
      url: 'https://old-icelandic.vercel.app/word/fadir',
      source: DictionarySource.OldIcelandic,
    },
    {
      url: 'https://old-swedish-dictionary.vercel.app/word/fadhir',
      source: DictionarySource.OldSwedish,
    },
  ]

  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(
      <Word
        entry={word}
        letters={getAlphabet()}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    )
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <Word
        entry={word}
        letters={getAlphabet()}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Returns null if entry is unavailable', () => {
    const tree = renderer.create(
      <Word
        entry={null}
        letters={getAlphabet()}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    ).toJSON()
    expect(tree).toBeNull()
  })

  test('Back button works', async () => {
    const tree = renderer.create(
      <Word
        entry={word}
        letters={getAlphabet()}
        abbreviations={abbreviations}
        crosslinks={crosslinks}
      />,
    )

    // Click back btn.
    await renderer.act(async () => {
      expect(mockHandler).not.toHaveBeenCalled()
      await tree.root.findByProps({ text: 'Back' }).props.action()

      // Assert mockrouter received a push.
      expect(mockHandler).toHaveBeenCalled()
      expect(mockHandler.mock.calls.length).toBe(1);
    })
  })
})

describe('Word page: data fetching', () => {
  test('getStaticPaths works', async () => {
    const expected = {
      paths: [],
      fallback: 'blocking',
    }

    const result = await getStaticPaths()

    expect(result).toMatchObject(expected)
  })

  test('getStaticProps works', async () => {
    const expected = {
      props: {
        entry: {
          word: 'þrályndi',
          partOfSpeech: 'n',
          definition: 'þrályndi, n. d. s.; hélt honum aptr fornsiðvenja ok mykyt þrályndi Barl. 12524;Fgrdl. 5816.',
          slug: 'thralyndi-2',
        },
        abbreviations: [
          {
            abbreviation: 'n.',
            explanation: 'Neutrum.',
          },
          {
            abbreviation: 's.',
            explanation: 'substantiv.',
          },
        ],
        crosslinks: [],
        letters: getAlphabet(),
      },
    }

    const result = await getStaticProps({ params: { word: 'thralyndi-2' } })

    expect(result).toEqual(expected)
  })

  test('getStaticProps returns 404 redirect for unkown words', async () => {
    const expected = {
      props: {},
      notFound: true,
    }

    const result = await getStaticProps({ params: { word: 'loremipsum' } })

    expect(result).toEqual(expected)
  })
})
