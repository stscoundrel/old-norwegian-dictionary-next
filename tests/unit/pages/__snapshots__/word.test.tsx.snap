// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word page: render & usage Matches snapshot 1`] = `
[
  <header>
    <div
      className="topBarSpacer"
    />
    <nav
      className="section"
    >
      <div
        className="topbar container"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <img
            alt="To home"
            data-nimg="1"
            decoding="async"
            height={30}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Ffavicon-48x48.png&w=64&q=75"
            srcSet="/_next/image?url=%2Ffavicon-48x48.png&w=32&q=75 1x, /_next/image?url=%2Ffavicon-48x48.png&w=64&q=75 2x"
            style={
              {
                "color": "transparent",
              }
            }
            width={30}
          />
        </a>
        <div
          aria-label="Open menu"
          className="hamburger"
          onClick={[Function]}
          role="button"
        >
          <span
            className="part"
          />
          <span
            className="part"
          />
          <span
            className="part"
          />
        </div>
      </div>
      <div
        className="content null container"
      >
        <ul
          className="list"
        >
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              a
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              á
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/b"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              b
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/d"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              d
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              e
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              é
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/f"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              f
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/g"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              g
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/h"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              h
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              i
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              í
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/j"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              j
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/k"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              k
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/l"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              l
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/m"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              m
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/n"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              n
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              o
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ó
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/p"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              p
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/r"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              r
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/s"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              s
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/t"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              t
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              u
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ú
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/v"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              v
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              y
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ý
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/th"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              þ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/ae"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              æ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe3"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              œ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ö
            </a>
          </li>
        </ul>
        <form
          className="form"
          onSubmit={[Function]}
        >
          <input
            aria-label="Search"
            className="input"
            name="search"
            onChange={[Function]}
            type="search"
            value=""
          />
          <button
            className="button button--secondary button--no-margs"
            name="submit"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </nav>
    <nav
      className="section"
    >
      <div
        className="container"
      >
        <a
          className="link"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Old Norwegian Dictionary
        </a>
        <a
          className="link"
          href="/letter/th"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Letter Þ
        </a>
        <a
          className="link"
          href="/word/thralyndi"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Þrályndi
        </a>
        <script
          dangerouslySetInnerHTML={
            {
              "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Old Norwegian Dictionary","item":"undefined/"},{"@type":"ListItem","position":2,"name":"Letter Þ","item":"undefined/letter/th"},{"@type":"ListItem","position":3,"name":"Þrályndi","item":"undefined/word/thralyndi"}]}",
            }
          }
          type="application/ld+json"
        />
      </div>
    </nav>
  </header>,
  <main
    className="container"
  >
    <article
      className="section"
    >
      <header>
        <h1
          lang="non"
        >
          Þrályndi
        </h1>
        <small
          className="subHeading"
        >
          Old Norwegian Dictionary - 
          þrályndi
        </small>
        <p>
          Meaning of Old Norwegian word 
          <em>
            "
            þrályndi
            "
          </em>
           in Norwegian.
        </p>
        <p>
          As defined by the Johan Fritzer's Old Norwegian dictionary:
        </p>
      </header>
      <dl
        className="definitionList"
      >
        <dt>
          <strong>
            þrályndi
          </strong>
           
        </dt>
        <dd
          className="itemDescription"
          dangerouslySetInnerHTML={
            {
              "__html": "þrályndi, n. d. s.; hélt honum aptr fornsiðvenja ok mykyt þrályndi Bar<abbr title="Linje.">L.</abbr> 12524;Fgrd<abbr title="Linje.">L.</abbr> 5816.",
            }
          }
          lang="nor"
        />
      </dl>
      <p>
        <strong>
          Part of speech:
        </strong>
         
        <em>
          n
        </em>
      </p>
      <p>
        Possible runic inscription in 
        <em>
          Medieval Futhork:
        </em>
        <span
          className="rune"
        >
          ᚦᚱᛆᛚᛦᚿᚦᛁ
        </span>
        <br />
        <small>
          Medieval Runes were used in Norway from 11th to 15th centuries.
        </small>
        <br />
        <small>
          Futhork was a continuation of earlier Younger Futhark runes, which were used to write Old Norse.
        </small>
      </p>
      <div
        className="abbreviations"
      >
        <div
          className="column"
        >
          <h4>
            Abbreviations used:
          </h4>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                f.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              Feminin.
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                L.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              Linje.
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                m.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              Masculin.
            </dd>
          </dl>
        </div>
      </div>
      <br />
      <h4>
        Also available in related dictionaries:
      </h4>
      <p>
        This headword also appears in dictionaries of other languages related to Old Norwegian.
      </p>
      <ul
        className="list"
      >
        <li
          className="listItem"
        >
          <a
            href="https://old-icelandic.vercel.app/word/fadir"
            rel="noopener noreferrer"
            target="_blank"
          >
            Old Icelandic - Geir Zoëga's Dictionary
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="https://old-swedish-dictionary.vercel.app/word/fadhir"
            rel="noopener noreferrer"
            target="_blank"
          >
            Old Swedish - K.F Söderwall's Dictionary
          </a>
        </li>
      </ul>
    </article>
    <a
      className="button"
      onClick={[Function]}
    >
      Back
    </a>
    <div
      aria-label="Back to top"
      className="backToTop"
      onClick={[Function]}
      role="button"
    >
      ↑
    </div>
  </main>,
  <footer
    className="section"
  >
    <div
      className="container"
    >
      <section
        className="section"
      >
        <h2>
          About
        </h2>
        <p>
          Based on "Dictionary of the Old Norwegian Language."
        </p>
        <p>
          Its is the largest Old Norse to Norwegian dictionary, containing over 40 000 word definitions. While the original dictionary is called dictionary of "Old Norwegian", it is practically a dictionary of western Old Norse. Technically "Old Norwegian" would be a later stage in the language.
        </p>
        <p>
          It was published in 1800s, which leads to there being public domain versions of the book available.
        </p>
      </section>
      <section
        className="section"
      >
        <h3
          className="h4"
        >
          Old Norwegian language
        </h3>
        <p>
          Old Norwegian is a early form of Norwegian language that was used from 11th to 14th century. It developed from Old West Norse.
        </p>
        <p>
          Also known as "Norwegian Norse", or "gammelnorsk" in modern Norwegian
        </p>
      </section>
      <section
        className="section"
      >
        <h4>
          Old Norse language
        </h4>
        <p>
          Old Norse was a North Germanic language that was spoken by inhabitants of Scandinavia and their overseas settlements from about the 7th to the 15th centuries.
        </p>
        <p>
          Also known as "the viking language", "Old Nordic", or "Old Scandinavian"
        </p>
      </section>
      <div
        className="navs"
      >
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Dictionary project
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norwegian-dictionary-next"
                rel="noopener noreferrer"
                target="_blank"
              >
                Source code
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norwegian-dictionary"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norwegian-dictionary-abbreviations"
                rel="noopener noreferrer"
                target="_blank"
              >
                Abbreviations
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/futhork"
                rel="noopener noreferrer"
                target="_blank"
              >
                Medieval Futhork Runes
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Related dictionary projects
          </h4>
          <ul>
            <li>
              <a
                href="https://old-swedish-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                K.F Söderwall's Old Swedish Dictionary
              </a>
            </li>
            <li>
              <a
                href="https://cleasby-vigfusson-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Cleasby and Vigfusson Old Norse Dictionary
              </a>
            </li>
            <li>
              <a
                href="https://old-icelandic.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                A Concise Dictionary of Old Icelandic
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norse-alphabet"
                rel="noopener noreferrer"
                target="_blank"
              >
                Old Norse Alphabet
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Quick links
          </h4>
          <ul
            className="navColumns"
          >
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                a
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                á
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/b"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                b
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/d"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                d
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                e
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                é
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/f"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                f
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/g"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                g
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/h"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                h
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                i
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                í
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/j"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                j
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/k"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                k
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/l"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                l
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/m"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                m
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/n"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                n
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                o
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ó
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/p"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                p
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/r"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                r
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/s"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                s
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/t"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                t
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                u
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ú
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/v"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                v
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                y
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ý
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/th"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                þ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/ae"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                æ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe3"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                œ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ö
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <small
        className="copyright"
      >
        Copyright © 2022 - 2021
        <br />
        <a
          href="https://github.com/stscoundrel"
          rel="noopener noreferrer"
          target="_blank"
        >
          Sampo Silvennoinen / StScoundrel
        </a>
      </small>
    </div>
  </footer>,
]
`;
